clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -3515,4 +3668,198 @@ void handleFailure(Throwable throwable) {
+            for (PartitionMetadata partitionMetadata : topicMetadata.partitionMetadata()) {
+                leaderEpochs.put(new TopicPartition(topicMetadata.topic(), partitionMetadata.partition()), partitionMetadata.leaderEpoch());
+            }
+        }
+        for (Map.Entry<TopicPartition, OffsetSpec> entry: topicPartitionOffsets.entrySet()) {
+
+            OffsetSpec offsetSpec = entry.getValue();
+            long offsetQuery = (offsetSpec instanceof TimestampSpec)
+                    ? ((TimestampSpec)offsetSpec).timestamp()
+                    : (offsetSpec instanceof OffsetSpec.EarliestSpec)
+                        ? ListOffsetRequest.EARLIEST_TIMESTAMP
+                        : ListOffsetRequest.LATEST_TIMESTAMP;
+            // avoid sending listOffsets request for topics with errors
+            if (!context.getErrors().containsKey(entry.getKey().topic())) {
+                Node node = context.getCluster().leaderFor(entry.getKey());
+                if (node != null) {
+                    if (!leaders.containsKey(node))
+                        leaders.put(node, new HashMap<>());
+                    leaders.get(node).put(entry.getKey(), new ListOffsetRequest.PartitionData(offsetQuery, leaderEpochs.get(entry.getKey())));
+                } else {
+                    KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(entry.getKey());
+                    future.completeExceptionally(Errors.LEADER_NOT_AVAILABLE.exception());
+                }
+            }
+        }
+
+        for (final Map.Entry<Node, Map<TopicPartition, ListOffsetRequest.PartitionData>> entry: leaders.entrySet()) {
+
+            final int brokerId = entry.getKey().id();
+
+            calls.add(new Call("listOffsets on broker " + brokerId, context.getDeadline(), new ConstantNodeIdProvider(brokerId)) {
+
+                @Override
+                ListOffsetRequest.Builder createRequest(int timeoutMs) {
+                    return ListOffsetRequest.Builder
+                            .forConsumer(true, context.getOptions().isolationLevel())
+                            .setTargetTimes(entry.getValue());
+                }
+
+                @Override
+                void handleResponse(AbstractResponse abstractResponse) {
+                    ListOffsetResponse response = (ListOffsetResponse) abstractResponse;
+                    for (Entry<TopicPartition, PartitionData> result : response.responseData().entrySet()) {
+
+                        KafkaFutureImpl<ListOffsetsResultInfo> future = futures.get(result.getKey());
+                        PartitionData partitionData = result.getValue();
+                        if (partitionData.error == Errors.NONE) {
+                            future.complete(new ListOffsetsResultInfo(partitionData.offset, partitionData.timestamp, partitionData.leaderEpoch));
+                        } else {
+                            future.completeExceptionally(result.getValue().error.exception());
```

Comment:
We would want to check for `InvalidMetadataException` here as well. We need to go back to the Metadata call if we find a metadata error. This is similar to the cases when we call `rescheduleFindCoordinatorTask`.